### PR TITLE
adding kvtag_scale_add_get test case for Metadata Retrieval Query test

### PR DESCRIPTION
**Pull request by zhangwei217245**
Difference between this one and the original kvtag_add_get_scale.c 
1. Now, multiple iterations of test can be performed.
2. In each iteration, we can now specify 
   1. the number of objects to be created, 
   2. the number of kvtags for each object.
   3. the length of attribute value of each kv tag
   4. the number of queries during each iteration
3. Now we can show how the throughput changes as 
   1. the number of objects increases in each iteration
   2. the number of attributes increases
   3. the length of attribute values increases
   4. the number of queries increases

Code tested on Cori with 8 nodes and 8*31 clients, and it passes the test. 

Here is the output from a smaller scale of test:
```
Iteration 1 : Objects: 1024 , Time: 0.3684 sec. Object throughput in this iteration: 2779.4284 .
Overall   1 : Objects: 1024 , Time: 0.3684 sec. Overall object throughput:           2779.4284 .
Iteration 1 : Tags: 3072 , Time: 0.0076 sec. Tag throughput in this iteration: 404206.8541 .
Overall   1 : Tags: 3072 , Time: 0.0076 sec. Overall tag throughput:           404206.8541 .
Iteration 1 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 77912.7678 .
Overall   1 : Queries: 30 , Time: 0.0004 sec. Overall query throughput:           77912.7678 .
starting creating 4 objects... 
Iteration 2 : Objects: 1024 , Time: 0.1400 sec. Object throughput in this iteration: 7313.3960 .
Overall   2 : Objects: 2048 , Time: 0.5084 sec. Overall object throughput:           4028.0222 .
Iteration 2 : Tags: 3072 , Time: 0.0076 sec. Tag throughput in this iteration: 402050.1088 .
Overall   2 : Tags: 6144 , Time: 0.0152 sec. Overall tag throughput:           403125.5968 .
Iteration 2 : Queries: 30 , Time: 0.0003 sec. Query throughput in this iteration: 86718.8973 .
Overall   2 : Queries: 60 , Time: 0.0007 sec. Overall query throughput:           82080.3131 .
starting creating 4 objects... 
Iteration 3 : Objects: 1024 , Time: 0.1592 sec. Object throughput in this iteration: 6431.5174 .
Overall   3 : Objects: 3072 , Time: 0.6677 sec. Overall object throughput:           4601.1854 .
Iteration 3 : Tags: 3072 , Time: 0.0072 sec. Tag throughput in this iteration: 427260.7318 .
Overall   3 : Tags: 9216 , Time: 0.0224 sec. Overall tag throughput:           410861.8616 .
Iteration 3 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 77338.1192 .
Overall   3 : Queries: 90 , Time: 0.0011 sec. Overall query throughput:           80436.2583 .
starting creating 4 objects... 
Iteration 4 : Objects: 1024 , Time: 0.1637 sec. Object throughput in this iteration: 6255.2319 .
Overall   4 : Objects: 4096 , Time: 0.8314 sec. Overall object throughput:           4926.8846 .
Iteration 4 : Tags: 3072 , Time: 0.0076 sec. Tag throughput in this iteration: 403359.0624 .
Overall   4 : Tags: 12288 , Time: 0.0300 sec. Overall tag throughput:           408960.1158 .
Iteration 4 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 68646.5466 .
Overall   4 : Queries: 120 , Time: 0.0016 sec. Overall query throughput:           77124.8054 .
starting creating 4 objects... 
Iteration 5 : Objects: 1024 , Time: 0.0026 sec. Object throughput in this iteration: 389072.1348 .
Overall   5 : Objects: 5120 , Time: 0.8340 sec. Overall object throughput:           6139.1704 .
Iteration 5 : Tags: 3072 , Time: 0.0067 sec. Tag throughput in this iteration: 460503.9989 .
Overall   5 : Tags: 15360 , Time: 0.0367 sec. Overall tag throughput:           418324.6720 .
Iteration 5 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 70256.3484 .
Overall   5 : Queries: 150 , Time: 0.0020 sec. Overall query throughput:           75645.7376 .
starting creating 4 objects... 
Iteration 6 : Objects: 1024 , Time: 0.0025 sec. Object throughput in this iteration: 413733.4839 .
Overall   6 : Objects: 6144 , Time: 0.8365 sec. Overall object throughput:           7345.2062 .
Iteration 6 : Tags: 3072 , Time: 0.0065 sec. Tag throughput in this iteration: 469224.3950 .
Overall   6 : Tags: 18432 , Time: 0.0433 sec. Overall tag throughput:           426026.9766 .
Iteration 6 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 73541.2741 .
Overall   6 : Queries: 180 , Time: 0.0024 sec. Overall query throughput:           75286.6693 .
starting creating 4 objects... 
Iteration 7 : Objects: 1024 , Time: 0.0025 sec. Object throughput in this iteration: 415414.1886 .
Overall   7 : Objects: 7168 , Time: 0.8389 sec. Overall object throughput:           8544.2279 .
Iteration 7 : Tags: 3072 , Time: 0.0066 sec. Tag throughput in this iteration: 465595.9344 .
Overall   7 : Tags: 21504 , Time: 0.0499 sec. Overall tag throughput:           431262.8537 .
Iteration 7 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 69750.0665 .
Overall   7 : Queries: 210 , Time: 0.0028 sec. Overall query throughput:           74442.5152 .
starting creating 4 objects... 
Iteration 8 : Objects: 1024 , Time: 0.0029 sec. Object throughput in this iteration: 352248.6095 .
Overall   8 : Objects: 8192 , Time: 0.8418 sec. Overall object throughput:           9731.1118 .
Iteration 8 : Tags: 3072 , Time: 0.0063 sec. Tag throughput in this iteration: 484704.5814 .
Overall   8 : Tags: 24576 , Time: 0.0562 sec. Overall tag throughput:           437289.5946 .
Iteration 8 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 76121.6697 .
Overall   8 : Queries: 240 , Time: 0.0032 sec. Overall query throughput:           74648.3471 .
starting creating 4 objects... 
Iteration 9 : Objects: 1024 , Time: 0.0024 sec. Object throughput in this iteration: 434273.7407 .
Overall   9 : Objects: 9216 , Time: 0.8442 sec. Overall object throughput:           10916.9228 .
Iteration 9 : Tags: 3072 , Time: 0.0062 sec. Tag throughput in this iteration: 491677.5505 .
Overall   9 : Tags: 27648 , Time: 0.0624 sec. Overall tag throughput:           442731.1103 .
Iteration 9 : Queries: 30 , Time: 0.0003 sec. Query throughput in this iteration: 86243.3996 .
Overall   9 : Queries: 270 , Time: 0.0036 sec. Overall query throughput:           75780.3854 .
starting creating 4 objects... 
Iteration 10 : Objects: 1024 , Time: 0.0023 sec. Object throughput in this iteration: 437235.8033 .
Overall   10 : Objects: 10240 , Time: 0.8465 sec. Overall object throughput:           12096.3562 .
Iteration 10 : Tags: 3072 , Time: 0.0063 sec. Tag throughput in this iteration: 488471.5251 .
Overall   10 : Tags: 30720 , Time: 0.0687 sec. Overall tag throughput:           446916.0266 .
Iteration 10 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 68985.2632 .
Overall   10 : Queries: 300 , Time: 0.0040 sec. Overall query throughput:           75041.2214 .
starting creating 4 objects... 
Iteration 11 : Objects: 1024 , Time: 0.0024 sec. Object throughput in this iteration: 420374.6008 .
Overall   11 : Objects: 11264 , Time: 0.8490 sec. Overall object throughput:           13267.8134 .
Iteration 11 : Tags: 3072 , Time: 0.0063 sec. Tag throughput in this iteration: 484085.4299 .
Overall   11 : Tags: 33792 , Time: 0.0751 sec. Overall tag throughput:           450057.5401 .
Iteration 11 : Queries: 30 , Time: 0.0003 sec. Query throughput in this iteration: 98380.8600 .
Overall   11 : Queries: 330 , Time: 0.0043 sec. Overall query throughput:           76695.3133 .
starting creating 4 objects... 
Iteration 12 : Objects: 1024 , Time: 0.0029 sec. Object throughput in this iteration: 348419.5097 .
Overall   12 : Objects: 12288 , Time: 0.8519 sec. Overall object throughput:           14424.0448 .
Iteration 12 : Tags: 3072 , Time: 0.0063 sec. Tag throughput in this iteration: 487326.0926 .
Overall   12 : Tags: 36864 , Time: 0.0814 sec. Overall tag throughput:           452944.1378 .
Iteration 12 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 79790.1839 .
Overall   12 : Queries: 360 , Time: 0.0047 sec. Overall query throughput:           76944.0196 .
starting creating 4 objects... 
Iteration 13 : Objects: 1024 , Time: 0.0024 sec. Object throughput in this iteration: 430443.7058 .
Overall   13 : Objects: 13312 , Time: 0.8543 sec. Overall object throughput:           15582.5347 .
Iteration 13 : Tags: 3072 , Time: 0.0062 sec. Tag throughput in this iteration: 493806.8405 .
Overall   13 : Tags: 39936 , Time: 0.0876 sec. Overall tag throughput:           455845.7848 .
Iteration 13 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 77101.1765 .
Overall   13 : Queries: 390 , Time: 0.0051 sec. Overall query throughput:           76956.0858 .
starting creating 4 objects... 
Iteration 14 : Objects: 1024 , Time: 0.0024 sec. Object throughput in this iteration: 435021.5027 .
Overall   14 : Objects: 14336 , Time: 0.8566 sec. Overall object throughput:           16735.0795 .
Iteration 14 : Tags: 3072 , Time: 0.0063 sec. Tag throughput in this iteration: 485691.2016 .
Overall   14 : Tags: 43008 , Time: 0.0939 sec. Overall tag throughput:           457855.4224 .
Iteration 14 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 77528.7246 .
Overall   14 : Queries: 420 , Time: 0.0055 sec. Overall query throughput:           76996.7079 .
starting creating 4 objects... 
Iteration 15 : Objects: 1024 , Time: 0.0024 sec. Object throughput in this iteration: 425244.2867 .
Overall   15 : Objects: 15360 , Time: 0.8591 sec. Overall object throughput:           17880.1810 .
Iteration 15 : Tags: 3072 , Time: 0.0064 sec. Tag throughput in this iteration: 481498.5758 .
Overall   15 : Tags: 46080 , Time: 0.1003 sec. Overall tag throughput:           459359.1581 .
Iteration 15 : Queries: 30 , Time: 0.0004 sec. Query throughput in this iteration: 69136.8791 .
Overall   15 : Queries: 450 , Time: 0.0059 sec. Overall query throughput:           76417.5392 .
starting creating 4 objects... 
Iteration 16 : Objects: 1024 , Time: 0.0029 sec. Object throughput in this iteration: 359020.9225 .
Overall   16 : Objects: 16384 , Time: 0.8619 sec. Overall object throughput:           19009.0797 .
Iteration 16 : Tags: 3072 , Time: 0.0063 sec. Tag throughput in this iteration: 485691.2016 .
Overall   16 : Tags: 49152 , Time: 0.1066 sec. Overall tag throughput:           460920.9775 .
Iteration 16 : Queries: 30 , Time: 0.0003 sec. Query throughput in this iteration: 89558.0925 .
Overall   16 : Queries: 480 , Time: 0.0062 sec. Overall query throughput:           77124.8054 .
Final Report: 
Servers: 8 , Clients: 248 , C/S ratio: 31.0000 
Iterations: 17 ,  Objects: 16384 , Tags/Object: 3 ,  Queries/Iteration: 10 , 
Object throughput: 19009.0797 , Tag Throughput: 460920.9775 , Query Throughput: 231374.4162 
```

### CODE REVIEW

**Comment by houjun** [comment]
@zhangwei217245  I'd suggest changing the name for the new source code file from "kvtag_scale_add_get.c" to something like "kvtag_add_get_benchmark.c".

**Comment by houjun** [comment]
@zhangwei217245 For the output, I'd suggest using a table-like CSV format with a header, so it is easier to parse for performance comparison. 
e.g. 
Iter, Type, Number of operations, Time (s), Throughput (per sec)
1, single create, 1024, 0.3684,  2779.4284
1, overall create,1024, 0.3684,  2779.4284
1, single add tag, 3072, 0.0076,  404206.8541
1, overall add tag, 3072, 0.0076,  404206.8541
1, single query tag, 30, 0.0004,  77912.7678 
1, overall query tag, 30, 0.0004,  77912.7678 
2, ...

**Comment by houjun** [comment]
Also, please install clang-format and do "make format" before you commit.

**Comment by zhangwei217245** [comment]
> @zhangwei217245 For the output, I'd suggest using a table-like CSV format with a header, so it is easier to parse for performance comparison.
> e.g.
> Iter, Type, Number of operations, Time (s), Throughput (per sec)
> 1, single create, 1024, 0.3684, 2779.4284
> 1, overall create,1024, 0.3684, 2779.4284
> 1, single add tag, 3072, 0.0076, 404206.8541
> 1, overall add tag, 3072, 0.0076, 404206.8541
> 1, single query tag, 30, 0.0004, 77912.7678
> 1, overall query tag, 30, 0.0004, 77912.7678
> 2, ...

As we discussed, I will write a bash script to convert the output to CSV. But to make sure the output is human readable as well, I prefer to reserve the current formatting.

**Comment by zhangwei217245** [comment]
> Also, please install clang-format and do "make format" before you commit.

addressed.

**Comment by zhangwei217245** [comment]
> @zhangwei217245 I'd suggest changing the name for the new source code file from "kvtag_scale_add_get.c" to something like "kvtag_add_get_benchmark.c".

addressed.

File: src/tests/CMakeLists.txt [hunk 1]
```diff
@@ -63,9 +63,11 @@ set(PROGRAMS
 #  data_server_meta_test
  kvtag_add_get
 #  kvtag_get
+ kvtag_add_get_benchmark
  kvtag_add_get_scale
 #  kvtag_query
-#  kvtag_query_scale
+ kvtag_query_scale
+ kvtag_scale_add_get
```
Comment (houjun) [inline src/tests/CMakeLists.txt]: remove this?

Comment (zhangwei217245) [reply src/tests/CMakeLists.txt]: > remove this?

addressed

**Review by houjun — APPROVED** [review]

**Pull request merged**